File: sdk/python/perpetual/resources/memory.py
from __future__ import annotations
from typing import Any
from urllib.parse import quote
from ..http import HttpClient
from ..types import MemoryEntry


class MemoryResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def set(
        self,
        namespace: str,
        key: str,
        value: Any,
        ttl_secs: int | None = None,
    ) -> dict:
        body: dict[str, Any] = {"value": value}
        if ttl_secs is not None:
            body["ttl_secs"] = ttl_secs
        return self._http.put(f"/v1/memory/{_enc(namespace)}/{_enc(key)}", body)

    def get(self, namespace: str, key: str) -> MemoryEntry:
        data = self._http.get(f"/v1/memory/{_enc(namespace)}/{_enc(key)}")
        return MemoryEntry(**data)

    def list(self, namespace: str) -> dict:
        return self._http.get(f"/v1/memory/{_enc(namespace)}")

    def delete(self, namespace: str, key: str) -> dict:
        return self._http.delete(f"/v1/memory/{_enc(namespace)}/{_enc(key)}")

    def clear(self, namespace: str) -> dict:
        return self._http.delete(f"/v1/memory/{_enc(namespace)}")

    def log_event(self, namespace: str, event_type: str, payload: Any) -> dict:
        return self._http.post(
            f"/v1/memory/{_enc(namespace)}/events",
            {"event_type": event_type, "payload": payload},
        )

    def events(
        self,
        namespace: str,
        limit: int = 50,
        before: str | None = None,
    ) -> dict:
        params = f"?limit={limit}"
        if before:
            params += f"&before={quote(before)}"
        return self._http.get(f"/v1/memory/{_enc(namespace)}/events{params}")


def _enc(s: str) -> str:
    return quote(s, safe="")

File: sdk/node/src/resources/memory.ts
import type { HttpClient } from '../client.js';
import type { MemoryEntry, MemoryKeyList } from '../types.js';

export class MemoryResource {
  constructor(private readonly http: HttpClient) {}

  set(namespace: string, key: string, value: unknown, opts?: { ttl_secs?: number }): Promise<{ id: string; namespace: string; key: string; expires_at: string | null }> {
    return this.http.put(`/v1/memory/${enc(namespace)}/${enc(key)}`, { value, ...opts });
  }

  get(namespace: string, key: string): Promise<MemoryEntry> {
    return this.http.get(`/v1/memory/${enc(namespace)}/${enc(key)}`);
  }

  list(namespace: string): Promise<MemoryKeyList> {
    return this.http.get(`/v1/memory/${enc(namespace)}`);
  }

  delete(namespace: string, key: string): Promise<{ deleted: boolean }> {
    return this.http.delete(`/v1/memory/${enc(namespace)}/${enc(key)}`);
  }

  clear(namespace: string): Promise<{ deleted_count: number }> {
    return this.http.delete(`/v1/memory/${enc(namespace)}`);
  }

  logEvent(namespace: string, eventType: string, payload: unknown): Promise<{ id: string; event_type: string; created_at: string }> {
    return this.http.post(`/v1/memory/${enc(namespace)}/events`, { event_type: eventType, payload });
  }

  events(namespace: string, opts?: { limit?: number; before?: string }): Promise<{ namespace: string; events: { id: string; event_type: string; payload: unknown; created_at: string }[]; has_more: boolean }> {
    const params = new URLSearchParams();
    if (opts?.limit) params.set('limit', String(opts.limit));
    if (opts?.before) params.set('before', opts.before);
    const qs = params.toString();
    return this.http.get(`/v1/memory/${enc(namespace)}/events${qs ? `?${qs}` : ''}`);
  }
}

function enc(s: string) { return encodeURIComponent(s); }
